# Ensemble Learning

## Initial Imports

In [ ]:
import warnings
warnings.filterwarnings('ignore')


In [2]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix
from imblearn.metrics import classification_report_imbalanced
from sklearn.preprocessing import LabelEncoder

## Read the CSV and Perform Basic Data Cleaning

In [4]:
# Load the data
file_path = Path('LoanStats_2019Q1.csv')
df = pd.read_csv(file_path)

le = LabelEncoder()

for col in df.select_dtypes(['object']):
    le.fit(df[col])
    df[col] = le.transform(df[col])


df.head()

loan_amnt  int_rate  installment  home_ownership  annual_inc  \
0    10500.0    0.1719       375.35               3     66000.0   
1    25000.0    0.2000       929.09               1    105000.0   
2    20000.0    0.2000       529.88               1     56000.0   
3    10000.0    0.1640       353.55               3     92000.0   
4    22000.0    0.1474       520.39               1     52000.0   

   verification_status  issue_d  loan_status  pymnt_plan    dti  ...  \
0                    1        2            1           0  27.24  ...   
1                    2        2            1           0  20.23  ...   
2                    2        2            1           0  24.26  ...   
3                    2        2            1           0  31.44  ...   
4                    0        2            1           0  18.76  ...   

   pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
0            85.7             100.0                   0.0        0.0   
1            91.2              50.0                   1.0        0.0   
2            66.7              50.0                   0.0        0.0   
3           100.0              50.0                   1.0        0.0   
4           100.0               0.0                   0.0        0.0   

   tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
0          65687.0            38199.0          2000.0   
1         271427.0            60641.0         41200.0   
2          60644.0            45684.0          7500.0   
3          99506.0            68784.0         19700.0   
4         219750.0            25919.0         27600.0   

   total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
0                     61987.0              0                     0  
1                     49197.0              0                     0  
2                     43144.0              0                     0  
3                     76506.0              0                     0  
4                     20000.0              0                     0  

[5 rows x 86 columns]

## Split the Data into Training and Testing

In [5]:
# Create our features
X = pd.get_dummies(df.drop(columns=["loan_status"]))

# Create our target
y = df.loan_status

In [6]:
X.describe()

loan_amnt      int_rate   installment  home_ownership    annual_inc  \
count  68817.000000  68817.000000  68817.000000    68817.000000  6.881700e+04   
mean   16677.594562      0.127718    480.652863        1.812779  8.821371e+04   
std    10277.348590      0.048130    288.062432        0.941313  1.155800e+05   
min     1000.000000      0.060000     30.890000        0.000000  4.000000e+01   
25%     9000.000000      0.088100    265.730000        1.000000  5.000000e+04   
50%    15000.000000      0.118000    404.560000        1.000000  7.300000e+04   
75%    24000.000000      0.155700    648.100000        3.000000  1.040000e+05   
max    40000.000000      0.308400   1676.230000        3.000000  8.797500e+06   

       verification_status       issue_d  pymnt_plan           dti  \
count         68817.000000  68817.000000     68817.0  68817.000000   
mean              0.669994      0.805542         0.0     21.778153   
std               0.719105      0.714932         0.0     20.199244   
min               0.000000      0.000000         0.0      0.000000   
25%               0.000000      0.000000         0.0     13.890000   
50%               1.000000      1.000000         0.0     19.760000   
75%               1.000000      1.000000         0.0     26.660000   
max               2.000000      2.000000         0.0    999.000000   

        delinq_2yrs  ...  pct_tl_nvr_dlq  percent_bc_gt_75  \
count  68817.000000  ...    68817.000000      68817.000000   
mean       0.217766  ...       95.057627         30.626217   
std        0.718367  ...        8.326426         33.631463   
min        0.000000  ...       20.000000          0.000000   
25%        0.000000  ...       93.000000          0.000000   
50%        0.000000  ...      100.000000         20.000000   
75%        0.000000  ...      100.000000         50.000000   
max       18.000000  ...      100.000000        100.000000   

       pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  total_bal_ex_mort  \
count          68817.000000    68817.0     6.881700e+04       6.881700e+04   
mean               0.125972        0.0     2.100332e+05       6.133843e+04   
std                0.336732        0.0     1.928088e+05       5.738798e+04   
min                0.000000        0.0     3.600000e+03       2.350000e+02   
25%                0.000000        0.0     6.697700e+04       2.650300e+04   
50%                0.000000        0.0     1.467100e+05       4.535700e+04   
75%                0.000000        0.0     3.036400e+05       7.657000e+04   
max                4.000000        0.0     3.292782e+06       1.295455e+06   

       total_bc_limit  total_il_high_credit_limit  hardship_flag  \
count    68817.000000                6.881700e+04        68817.0   
mean     29734.128558                5.572240e+04            0.0   
std      26795.394232                5.095845e+04            0.0   
min        100.000000                1.270000e+02            0.0   
25%      11600.000000                2.288000e+04            0.0   
50%      22100.000000                4.200000e+04            0.0   
75%      39300.000000                7.249900e+04            0.0   
max     509400.000000                1.426964e+06            0.0   

       debt_settlement_flag  
count               68817.0  
mean                    0.0  
std                     0.0  
min                     0.0  
25%                     0.0  
50%                     0.0  
75%                     0.0  
max                     0.0  

[8 rows x 85 columns]

In [7]:
# Check the balance of our target values
y[df.loan_status].value_counts()

1    68817
Name: loan_status, dtype: int64

In [8]:
# Split the X and y into X_train, X_test, y_train, y_test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)
X_train

loan_amnt  int_rate  installment  home_ownership  annual_inc  \
3903      1600.0    0.0881        50.74               2     35964.0   
28390     9000.0    0.0756       280.21               2     41000.0   
15470    10000.0    0.1033       214.10               1    112000.0   
4279     36000.0    0.1033      1167.21               3    120000.0   
57514    18000.0    0.0881       570.81               1     51000.0   
...          ...       ...          ...             ...         ...   
49100     8400.0    0.0646       257.30               1     37992.0   
20609    30000.0    0.0646       918.93               1    217000.0   
21440    20000.0    0.1180       442.88               1     35000.0   
50057    11625.0    0.1447       273.34               3     30000.0   
5192     20000.0    0.1033       428.20               1     60000.0   

       verification_status  issue_d  pymnt_plan    dti  delinq_2yrs  ...  \
3903                     0        2           0  24.12          0.0  ...   
28390                    0        0           0  16.89          1.0  ...   
15470                    2        0           0  17.75          0.0  ...   
4279                     0        2           0  19.95          0.0  ...   
57514                    2        1           0  19.11          1.0  ...   
...                    ...      ...         ...    ...          ...  ...   
49100                    0        1           0  33.16          0.0  ...   
20609                    0        0           0  19.56          0.0  ...   
21440                    0        0           0  16.19          0.0  ...   
50057                    0        1           0  18.96          0.0  ...   
5192                     1        2           0  59.56          0.0  ...   

       pct_tl_nvr_dlq  percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  \
3903            100.0               0.0                   0.0        0.0   
28390            95.0              33.3                   0.0        0.0   
15470            92.6              20.0                   0.0        0.0   
4279            100.0              40.0                   0.0        0.0   
57514            95.5              25.0                   0.0        0.0   
...               ...               ...                   ...        ...   
49100           100.0               0.0                   0.0        0.0   
20609           100.0               7.7                   0.0        0.0   
21440           100.0               0.0                   0.0        0.0   
50057            86.7              20.0                   0.0        0.0   
5192            100.0              20.0                   0.0        0.0   

       tot_hi_cred_lim  total_bal_ex_mort  total_bc_limit  \
3903          148200.0            14292.0         58300.0   
28390          68292.0            52587.0         24500.0   
15470         416023.0            78944.0         65400.0   
4279          116200.0            73826.0         85200.0   
57514         221075.0           126195.0         36500.0   
...                ...                ...             ...   
49100         119052.0            67055.0         52800.0   
20609         358467.0           206670.0        137300.0   
21440         140700.0             7641.0         13700.0   
50057          96441.0            79650.0         22700.0   
5192          477516.0           200731.0         62100.0   

       total_il_high_credit_limit  hardship_flag  debt_settlement_flag  
3903                      15000.0              0                     0  
28390                     40492.0              0                     0  
15470                     80642.0              0                     0  
4279                      30000.0              0                     0  
57514                     96715.0              0                     0  
...                           ...            ...                   ...  
49100                     66252.0              0                     0  
20609      

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [9]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [10]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
X_scaler = scaler.fit(X_train)

In [11]:
# Scale the training and testing data
X_test_scaled = X_scaler.transform(X_test)
X_train_scaled = X_scaler.transform(X_train)

## Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Display the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier only, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [12]:
# Resample the training data with the BalancedRandomForestClassifier
from imblearn.ensemble import BalancedRandomForestClassifier
brf = BalancedRandomForestClassifier(n_estimators=100, random_state=1)
brf.fit(X_train, y_train)

BalancedRandomForestClassifier(random_state=1)

In [13]:
# Calculated the balanced accuracy score
y_pred = brf.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.7290249400290825

In [14]:
# Display the confusion matrix
confusion_matrix(y_test, y_pred)

array([[   62,    39],
       [ 2665, 14439]], dtype=int64)

In [15]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.02      0.61      0.84      0.04      0.72      0.51       101
          1       1.00      0.84      0.61      0.91      0.72      0.53     17104

avg / total       0.99      0.84      0.62      0.91      0.72      0.53     17205



In [16]:
# List the features sorted in descending order by feature importance
importances = brf.feature_importances_

importances_sorted = sorted(zip(brf.feature_importances_, X.columns), reverse=True)
importances_sorted[:20]

[(0.07646711017201446, 'total_rec_prncp'),
 (0.0681731200022894, 'last_pymnt_amnt'),
 (0.05699252404641979, 'total_pymnt'),
 (0.05424778840861491, 'total_pymnt_inv'),
 (0.04735017423439608, 'total_rec_int'),
 (0.031245651401774252, 'int_rate'),
 (0.022619745828795626, 'issue_d'),
 (0.02244519942392171, 'mths_since_recent_inq'),
 (0.020409215004023765, 'installment'),
 (0.01808337089191728, 'tot_hi_cred_lim'),
 (0.017004195413593538, 'bc_util'),
 (0.01640559494238341, 'dti'),
 (0.016118934293676203, 'mths_since_rcnt_il'),
 (0.01581716497633785, 'il_util'),
 (0.015630310867867408, 'max_bal_bc'),
 (0.015494921031886601, 'out_prncp'),
 (0.015281199871082264, 'out_prncp_inv'),
 (0.01470961095035287, 'mo_sin_old_il_acct'),
 (0.014693851231416522, 'mo_sin_old_rev_tl_op'),
 (0.01425370693271944, 'total_bal_ex_mort')]

### Easy Ensemble Classifier

In [17]:
# Train the Classifier
from imblearn.ensemble import EasyEnsembleClassifier
ee = EasyEnsembleClassifier(n_estimators=100, random_state=1)
ee.fit(X_train, y_train)

EasyEnsembleClassifier(n_estimators=100, random_state=1)

In [18]:
# Calculated the balanced accuracy score
y_pred = ee.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9316600714093861

In [19]:
# Display the confusion matrix
confusion_matrix(y_test, y_pred)

array([[   93,     8],
       [  983, 16121]], dtype=int64)

In [20]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.09      0.92      0.94      0.16      0.93      0.87       101
          1       1.00      0.94      0.92      0.97      0.93      0.87     17104

avg / total       0.99      0.94      0.92      0.97      0.93      0.87     17205



### Final Questions

1. Which model had the best balanced accuracy score?
    The Easy Ensemble Classifier model had the best Balanced accuracy score - .93 vs .72
    
2. Which model had the best recall score?

    The Easy Ensemble Classifier model had the best recall score - .94 vs .84

3. Which model had the best geometric mean score?

    The Easy Ensemble Classifier model had the best geometric mean score - .93 vs .72

4. What are the top three features?

    YOUR ANSWER HERE.